### PR TITLE
RPC server for API Interface 

### PR DESCRIPTION
**Pull request by yamabiiko**
Opening this PR so it can start getting reviewed and possibly get some help.
As discussed in #1185, this is quite a substantial rework which adds the possibility of running an RPC server and unifies the RPC methods and CLI commands under an internal API.

To be done before we can get this merged:
- [x] Tests: fixing and expanding tests in `swap/tests/rpc.rs`: right now the test runs the RPC server but the functionality cannot be tested as the Bitcoin wallet does not have any balance and the swap database is empty (so right now it's using some placeholder BTC/XMR addresses and swap ids and it will always fail).
- [x] Tests: write more tests (containers?)
- [x] Reevaluate the placing of the new files (request.rs, api.rs, rpc.rs) within the project file strucuture
- [ ] Organize commits by squashing some, changing commit messages to be more descriptive to make reviewing this easier
- [ ] Get the entire code reviewed by one other person (@delta1)

If someone could help with writing proper tests (or pointing out at some examples in the code to how swaps and transactions can be simulated/tested) it would be appreciated.

### CODE REVIEW

**Comment by binarybaron** [comment]
@delta1 @yamabiiko Please give as much feedback and scrutiny as possible. The API is really important and we need to make sure that it's stable enough to be used by a lot of GUIs etc. in the future. I'll also give feedback in the next days and try to be as thorough as possible.

**Comment by delta1** [comment]
Will review when I can in the next few days. @yamabiiko is there a reason for duplicated code instead of importing from other modules?

**Comment by yamabiiko** [comment]
> Will review when I can in the next few days. @yamabiiko is there a reason for duplicated code instead of importing from other modules?

Could you point out the specific examples?

**Comment by yamabiiko** [comment]
Another thing that still needs to be addressed is printing the tables/vectors with the History and ListSellers commands (just like before with the CLI)

**Comment by binarybaron** [comment]
True + we need to be able to associate log messages to individual rpc requests/commands/swaps

File: swap/src/api.rs
Comment (binarybaron) [inline swap/src/api.rs:99]: Maybe rename this to `context`? I think you called this Init before but renamed the struct to Context. Up to you to decide

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:277]: We need to return the file locations as JSON here but you've already mentioned this in the PR description.

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:316]: Can we add an option to allow the user to set the daemon listen address manually?

**Review by binarybaron — COMMENTED** [review]
Left a few comments

File: swap/src/api/request.rs
Comment (yamabiiko) [reply swap/src/api/request.rs:316]: Yes of course. It's meant to be that way, I haven't changed from the default value back yet

**Comment by yamabiiko** [comment]
Could this be reviewed again? Specifically regarding CLI/Json output. The only thing I see that's missing right now is some tests for checking if concurrent swaps are correctly handled.

**Comment by binarybaron** [comment]
I'll take a look. Thanks!

**Comment by binarybaron** [comment]
> Could this be reviewed and tested again? Specifically regarding CLI/Json output. The only thing I see that's missing right now is some tests for checking if concurrent swaps are correctly handled.

Are concurrent swaps handled at all?

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:83]: Is there a better way to structure this?
We are making all fields optional here which requires us to later again .unwrap() to make sure they exist. Could we maybe combine the Method and Params struct to assign all the required fields to each Method individually?

@delta1 Perhaps you have some input

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:179]: Where are exceptions caught? Are they caught or will this cause the CLI to panic?

**Review by binarybaron — COMMENTED** [review]
Left some comments. Not finished with the review yet

File: swap/src/api/request.rs
Comment (yamabiiko) [reply swap/src/api/request.rs:83]: I'll play around with it in the next days to see if it can be structured in a cleaner way

File: swap/src/api/request.rs
Comment (yamabiiko) [reply swap/src/api/request.rs:179]: If an error gets caught, the function will[ return early with the error](https://docs.rs/anyhow/latest/anyhow/macro.bail.html). So it's not a panic.

Comment (binarybaron) [reply swap/src/api/request.rs:179]: Is the error returned to the API caller?

**Comment by binarybaron** [comment]
Here is a list of things that we still need to get done before we can get this merged. I'll also try to help where I can and make some commits. I'll update this list when needed.

- [ ] Currently we save the logs of each swap to an individual log file (swap-SWAPID.log). This allows other programs (e.g the GUI) to parse information from those log files about the state of the certain swaps. We need to continue this behaviour. I think this can be achieved by using [spans](https://docs.rs/tracing/latest/tracing/span/index.html#)
- [ ] It'd be best if we could associate each log message that is printed to the corresponding API command that caused it. Again, I believe this can be done using spans
- [ ] We need to be able to forcefully terminate a running swap without crashing the RPC server

**Comment by binarybaron** [comment]
> Here is a list of things that we still need to get done before we can get this merged. I'll also try to help where I can and make some commits. I'll update this list when needed.
> 
> * [ ]  Currently we save the logs of each swap to an individual log file (swap-SWAPID.log). This allows other programs (e.g the GUI) to parse information from those log files about the state of the certain swaps. We need to continue this behaviour. I think this can be achieved by using [spans](https://docs.rs/tracing/latest/tracing/span/index.html#)
> * [ ]  It'd be best if we could associate each log message that is printed to the corresponding API command that caused it. Again, I believe this can be done using spans
> * [ ]  We need to be able to forcefully terminate a running swap without crashing the RPC server

I made a commit to fix the second bullet point.

File: swap/src/api/request.rs
Comment (yamabiiko) [reply swap/src/api/request.rs:190]: Yes, the API methods returns a Result<value, error>

**Review by binarybaron — COMMENTED** [review]
Left some more comments

**Comment by yamabiiko** [comment]
Should the running swap be cancelled with a separate method? Or should we do it when  `CancelAndRefund` gets called?

**Comment by binarybaron** [comment]
Seperate method. CancelAndRefund just means that both the cancel and refund transactions are published.

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:227]: I suppose this was for testing?

**Review by binarybaron — CHANGES_REQUESTED** [review]

File: swap/src/api/request.rs
Comment (yamabiiko) [reply swap/src/api/request.rs:227]: Yes, accidentally kept this line

**Comment by binarybaron** [comment]
> Here is a list of things that we still need to get done before we can get this merged. I'll also try to help where I can and make some commits. I'll update this list when needed.
> 
> * [x]  Currently we save the logs of each swap to an individual log file (swap-SWAPID.log). This allows other programs (e.g the GUI) to parse information from those log files about the state of the certain swaps. We need to continue this behaviour. I think this can be achieved by using [spans](https://docs.rs/tracing/latest/tracing/span/index.html#) (discarded)
> * [x]  It'd be best if we could associate each log message that is printed to the corresponding API command that caused it. Again, I believe this can be done using spans
> * [x]  We need to be able to forcefully terminate a running swap without crashing the RPC server
> * [x]  We need to prevent the user from initiating multiple swaps concurrently
> * [ ]  We need to add an endpoint to retrieve information about the currently running swap
> * [x]  We should look into combining the Cmd and Params struct

This PR is getting closer and closer to being finished. We're mostly missing two things:
* [ ] The tests are not complete and not working atm
* [ ] We want to make it possible to send push notifications to API callers about the currently running swap. Add a callback(?) to the context to allow the BuyXmr/Resume method to push JSON RPC notifications to the API caller to notify them about the status of the swap

#### Due to the massive changes triggered by this PR and its sheer size, I'll make an special exception to the bounty system for this one. **All reviewers, excluding me, will receive a payout of `1 XMR` for conducting a thorough review of the code**. I want to make sure that we minimise the chance of breaking anything major here and I think the risk is quite high due to the large scale modifications to the whole codebase.

In practice, this will probably only apply to you @delta1 unless someone else with sufficient knowledge of rust and this codebase come forth as well.

**Comment by yamabiiko** [comment]
> > Here is a list of things that we still need to get done before we can get this merged. I'll also try to help where I can and make some commits. I'll update this list when needed.
> > 
> > * [x]  Currently we save the logs of each swap to an individual log file (swap-SWAPID.log). This allows other programs (e.g the GUI) to parse information from those log files about the state of the certain swaps. We need to continue this behaviour. I think this can be achieved by using [spans](https://docs.rs/tracing/latest/tracing/span/index.html#) (discarded)
> > * [x]  It'd be best if we could associate each log message that is printed to the corresponding API command that caused it. Again, I believe this can be done using spans
> > * [x]  We need to be able to forcefully terminate a running swap without crashing the RPC server
> > * [x]  We need to prevent the user from initiating multiple swaps concurrently
> > * [ ]  We need to add an endpoint to retrieve information about the currently running swap
> > * [x]  We should look into combining the Cmd and Params struct
> 
> This PR is getting closer and closer to being finished. We're mostly missing two things:
> * [ ] The tests are not complete and not working atm
> * [ ] We want to make it possible to send push notifications to API callers about the currently running swap. Add a callback(?) to the context to allow the BuyXmr/Resume method to push JSON RPC notifications to the API caller to notify them about the status of the swap
> * [ ] Reevaluate the placing of the new files (request.rs, api.rs, rpc.rs) within the project file strucuture
> 
> #### Due to the massive changes triggered by this PR and its sheer size, I'll make an special exception to the bounty system for this one. **All reviewers, excluding me, will receive a payout of `1 XMR` for conducting a thorough review of the code**. I want to make sure that we minimise the chance of breaking anything major here and I think the risk is quite high due to the large scale modifications to the whole codebase.
> 
> In practice, this will probably only apply to you @delta1 unless someone else with sufficient knowledge of rust and this codebase come forth as well.

I am working on the second point, when that will be done I will take care of tests as well (though if someone else wants to they can help).

I will also review everything when we have completed these points, but as binarybaron mentioned an external reviewer would be really beneficial to the codebase quality and reliability.

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:177]: This comment is outdated. Can be removed

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:186]: This if statement can be removed. The called function will already bail if no swap lock is active

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:536]: Perhaps this should be renamed to something more clear: GetRawState

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:714]: We should return those keys here

File: swap/src/bin/swap.rs
Comment (binarybaron) [inline swap/src/bin/swap.rs:24]: I believe this is unused

File: swap/src/bitcoin.rs
Comment (binarybaron) [inline swap/src/bitcoin.rs:297]: This needs testing.

**Review by binarybaron — CHANGES_REQUESTED** [review]
Left a few notes for me and @yamabiiko

**Comment by binarybaron** [comment]
@CodiumAI-Agent /review

**Comment by binarybaron** [comment]
> @CodiumAI-Agent /review

Just wanted to try this once. I'm not gonna put too much belief into whatever it produces 😁

**Comment by binarybaron** [comment]
I updated the PR description

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:534]: We always return "ok" here even if the swap does not exist or something goes wrong. I think its fine as is but if someone knows how to implement this such that we fail if we encounter an error in the very beginning, we could do that

File: swap/src/api/request.rs
Comment (binarybaron) [inline swap/src/api/request.rs:620]: I don't think we need to return the whole tx here. Amount and txid should be enough. Returning the fee as well could be useful

**Review by binarybaron — CHANGES_REQUESTED** [review]

**Comment by binarybaron** [comment]
@delta1 When you review this can you give a comment on the placing of the request.rs, api.rs, rpc.rs files within the project structure. I'm not too familiar with how this is typically done in rust projects.

File: swap/src/api.rs
Comment (delta1) [inline swap/src/api.rs:235]: not sure this is the best place for these methods

File: swap/src/monero.rs
Comment (delta1) [inline swap/src/monero.rs:50]: why is this necessary? it's a bit of a footgun to implement Display like this for secrets. What I've seen in a crate like rust-secp256k1 is that the Display impl still hides the secret, and it's necessary to call a very specific `unsafe_display_secret` or something if you want to view the actual secret.  

https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/secret.rs#L66

File: swap/src/rpc/methods.rs
Comment (delta1) [inline swap/src/rpc/methods.rs:22]: this method should return a Result so that these unwraps can just be `?`

**Review by delta1 — CHANGES_REQUESTED** [review]
Concept ACK, with the following caveats: 

- merge conflicts have to be resolved. either rebase this branch on master or just try merge master into this and resolve the conflicts 
- i think the init wallet methods that have been moved into api should probably be in their own module if possible. if not then it's probably fine. 

left some other small comments, when this branch is free of conflicts and passing CI then I'll do another pass.

File: otp
Comment (delta1) [reply otp:1]: removed

File: swap/src/monero.rs
Comment (binarybaron) [reply swap/src/monero.rs:50]: I did this so that we can format the private key properly to be printed to the console / API caller.

I'll modify it to use the display_secret method instead. I didn't know that existed, thanks :)

File: swap/src/api.rs
Comment (binarybaron) [reply swap/src/api.rs:235]: Agreed. Where'd you put it?

Comment (yamabiiko) [reply swap/src/api.rs:235]: Also this code duplicated in in `src/bin/asb.rs`

File: swap/src/rpc/methods.rs
Comment (yamabiiko) [reply swap/src/rpc/methods.rs:22]: The methods are inside `register_modules`, which returns a `RpcModule<Arc<Context>>`. Not sure if it would make sense to wrap it inside a Result here.

Comment (delta1) [reply swap/src/rpc/methods.rs:22]: you can just use a ? where that method is called.

File: swap/src/api.rs
Comment (delta1) [reply swap/src/api.rs:235]: i think it should go in a new init module where it can be called from both places

File: swap/src/api/request.rs
Comment (yamabiiko) [reply swap/src/api/request.rs:307]: It ensures that the check for swap force suspension is run first, which seems like a sensible behavior to me. Why do you think it can be removed?

File: swap/Cargo.toml
Comment (yamabiiko) [reply swap/Cargo.toml]: I removed jsonrpsee-types since we don't use it anymore, but here we are pulling the WebSocket client feature of jsonrpsee which we use for tests only

**Comment by binarybaron** [comment]
@yamabiiko the code you pushed for moving initialization out of the spawned task looks good to me! Can you implement the same for BuyXmr as well?

File: swap/src/api/request.rs
Comment (binarybaron) [reply swap/src/api/request.rs:307]: You're right. It doesn't hurt to run it first.

**Comment by yamabiiko** [comment]
> @yamabiiko the code you pushed for moving initialization out of the spawned task looks good to me! Can you implement the same for BuyXmr as well?

Yes, but I need to refactor the code to avoid borrowing the context before moving, I'll work on it

**Comment by binarybaron** [comment]
> > @yamabiiko the code you pushed for moving initialization out of the spawned task looks good to me! Can you implement the same for BuyXmr as well?
> 
> Yes, but I need to refactor the code to avoid borrowing the context before moving, I'll work on it

Yeah I had some issues with that too when I tried to implement this

File: swap/src/monero.rs
Comment (binarybaron) [reply swap/src/monero.rs:50]: I took a look again and I think I'm really just continuing how the display of secret was treated before.

```rust
let (spend_key, view_key) = state5.xmr_keys();

                    let address = monero::Address::standard(
                        context.config.env_config.monero_network,
                        monero::PublicKey::from_private_key(&spend_key),
                        monero::PublicKey::from(view_key.public()),
                    );

                    tracing::info!(address=%address, spend_key=%spend_key, view_key=view_key.to_secret_string(), "Monero recovery information");

                    Ok(json!({
                        "address": address,
                        "spend_key": spend_key.to_string(),
                        "view_key": view_key.to_secret_string(),
                    }))
```

There are two types of Monero keys. Spend key (which is of type PrivateKey) and ViewKey which is of type (PrivateViewKey). PrivateKey implements Display already. PrivateViewKey is basically just a wrapped ViewKey with some extra traits. The code I added just calls that underlying Display implementation.

Therefore, the current convention in this project seems to be implement display even for secrets. SpendKey is arguably more secret than ViewKey and already implements Display so why shouldn't ViewKey as well?

Do you agree or should I remove the Display implementation for all private key structs. Seems out of scope for this PR to me imo.

**Comment by binarybaron** [comment]
We broke rustfmt... :(



```bash
Message: Child process exited with code 1: error[internal]: left behind trailing whitespace
```

**Review by binarybaron — CHANGES_REQUESTED** [review]

**Comment by binarybaron** [comment]
This is ready for review @delta1

**Comment by delta1** [comment]
> This is ready for review @delta1

thanks i should be able to review again next week

**Comment by binarybaron** [comment]
bors try

**Comment by delta1** [comment]
resolved conflicts, running CI

**Comment by delta1** [comment]
pushed clippy and minor fixes 

will review further soon

**Comment by delta1** [comment]
I think the RPC test needs to be in CI and  modified before this can be merged. Since the tests need to run in serial, it might as well be one test that just runs through the various calls.

**Comment by yamabiiko** [comment]
How exactly do you think the tests need to be changed, apart from grouping together the serial ones?

**Comment by delta1** [comment]
> I think leaving the tests separated is fine to avoid waiting on one big slow test.

I disagree, especially since they don't work. 

Conflicts need to be resolved to run CI and then will re-review @yamabiiko

**Comment by yamabiiko** [comment]
> I disagree, especially since they don't work.

Could you elaborate more? Locally they work fine, and I don't see them fail in the CI.
It seems the Monero wallet RPC failed but that should be unrelated.

**Comment by yamabiiko** [comment]
> Locally they work fine

Ok, I somehow wasn't running all of them.
Now they should be fixed and merged into three tests, since some of them use the `XmrLockTransactionSent` state, one needs to mock a swap suspend and all the other have no running swap.

**Comment by yamabiiko** [comment]
CI can be re-run

**Comment by yamabiiko** [comment]
Seems like the ubuntu-latest runner is full. Maybe [this](https://github.com/marketplace/actions/free-disk-space-ubuntu) can be helpful

**Comment by delta1** [comment]
> Seems like the ubuntu-latest runner is full. Maybe [this](https://github.com/marketplace/actions/free-disk-space-ubuntu) can be helpful

testing on my fork here: https://github.com/delta1/xmr-btc-swap/actions/runs/7872376768/job/21477456268

**Comment by yamabiiko** [comment]
Seems like all the main concerns have been addressed and tests are passing successfully. Are we waiting for a final review or some other process I can contribute to?

**Comment by delta1** [comment]
> Seems like all the main concerns have been addressed and tests are passing successfully. Are we waiting for a final review or some other process I can contribute to?

indeed sorry i am busy with some personal issues but should have a final review/approval next week 

@binarybaron do you have a moment to review, and test this?

**Comment by binarybaron** [comment]
I will try to review it this weekend

File: swap/src/api/request.rs
Comment (delta1) [reply swap/src/api/request.rs]: @binarybaron was this addressed?

File: swap/src/bitcoin/wallet.rs
Comment (binarybaron) [reply swap/src/bitcoin/wallet.rs:196]: Yes

**Comment by binarybaron** [comment]
> @binarybaron was this addressed? 

Can you expand? Don't know what you mean

**Review by delta1 — APPROVED** [review]
I'm happy for this to go in with @binarybaron's approval.

**Comment by yamabiiko** [comment]
> @binarybaron was this addressed?

Yes, the swap lock release logic was addressed by https://github.com/comit-network/xmr-btc-swap/pull/1276/commits/c7f411c8c9ac0d4360a93a31fffdd28e2a2850ec and improved in the next few commits

**Comment by binarybaron** [comment]
I'll do a few last tests and then this can be merged.

@yamabiiko Can you give me an XMR address to send the Bounty to?

I want to thank you for all your work and I apologize for the long time it took to get this PR merged.

**Comment by delta1** [comment]
> I'll do a few last tests and then this can be merged. 

Sounds good, I just resolved the conflicts.

**Comment by delta1** [comment]
@binarybaron merging this now, any issues can be fixed in a follow-up. 

@yamabiiko congrats! thank you for the contribution :pray:

**Comment by binarybaron** [comment]
> @binarybaron merging this now, any issues can be fixed in a follow-up.
> 
> @yamabiiko congrats! thank you for the contribution 🙏

Thanks for merging :))) we're missing a changelog sadly but its no big deal

**Comment by delta1** [comment]
yes definitely need a changelog entry.

**Pull request merged**
